overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,232 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as pulumi from "@pulumi/pulumi"
+import { Bucket } from "./bucket";
+import { BucketNotification } from "./bucketNotification";
+import * as lambda from "../lambda"
+
+/**
+ * Arguments to help customize a notification subscription for a bucket.
+ */
+export interface CommonBucketSubscriptionArgs {
+    /**
+     * An optional prefix or suffix to filter down notifications.  See
+     * aws.s3.BucketNotification.lambdaFunctions for more details.
+     */
+    filterPrefix?: string;
+    filterSuffix?: string;
+}
+
+export interface BucketEventSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    /**
+     * Events to subscribe to. For example: "[s3:ObjectCreated:*]".  Cannot be empty.
+     */
+    events: string[];
+}
+
+/**
+ * Arguments to specifically control a subscription to 'ObjectCreated' notifications on a bucket.If
+ * more events than just 'ObjectCreated' events are desired, the 'subscribe' function should be used
+ * instead.
+ */
+export interface ObjectCreatedSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    // The type of event to listen for.  '*' is the default if not provided.
+    event?: "*" | "Put" | "Post" | "Copy" | "CompleteMultipartUpload";
+}
+
+/**
+ * Arguments to specifically control a subscription to 'ObjectRemoved' notifications on a bucket. If
+ * more events than just 'ObjectRemoved' events are desired, the 'subscribe' function should be used
+ * instead.
+ */
+export interface ObjectRemovedSubscriptionArgs extends CommonBucketSubscriptionArgs {
+    // The type of event to listen for.  '*' is the default if not provided.
+    event?: "*" | "Delete" | "DeleteMarkerCreated";
+}
+
+// See https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html.
+export interface BucketEvent {
+    Records?: BucketRecord[];
+}
+
+export interface BucketRecord {
```

Comment:
Nit, could use some better comments. I like where we landed on this general topic in https://github.com/pulumi/pulumi-aws/pull/323.

Reply:
A large problem here is trying to find acutal docs to refer to for these guys.  For example, this is the best i could find from amazon themselves: https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html

Note that htey don't really explain what the fields are in their own docs.  It's hard to give better comments here when AWS isn't even really documenting this stuff in any explicit fashion (that i can find at least).

Reply:
Perhaps just including that doc link in the interface docs would be nice.